### PR TITLE
Add noop_tracer to seperate the logic of handling tracing when not sampled

### PR DESCRIPTION
**Pull request by liyanhui1228**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -185,9 +185,10 @@ Customize the sampler, reporter, propagator in the ``settings.py`` file:
 ::
 
     OPENCENSUS_TRACE = {
-        'TRACER': 'opencensus.trace.tracer.context_tracer.ContextTracer',
-        'SAMPLER': 'opencensus.trace.samplers.always_on.AlwaysOnSampler',
-        'REPORTER': 'opencensus.trace.reporters.print_reporter.PrintReporter'
+        'SAMPLER': 'opencensus.trace.samplers.fixed_rate.FixedRateSampler',
+        'REPORTER': 'opencensus.trace.reporters.print_reporter.PrintReporter',
+        'PROPAGATOR': 'opencensus.trace.propagation.google_cloud_format.'
```
[the remark this answers is not in the document]
Comment (duggelz) [reply README.rst]: Also, can you make a diagram of the different classes on the whiteboard and how they interact?  I'm losing track.

File: trace/opencensus/trace/tracer/context_tracer.py [hunk 2]
```diff
@@ -194,6 +115,10 @@ def end_span(self):
         else:
             self.span_context.span_id = self._span_stack[-1].span_id
 
+    def current_span(self):
+        """Return the current span."""
+        return self._span_stack[-1]
```
Comment (duggelz) [inline trace/opencensus/trace/tracer/context_tracer.py]: Can `_span_stack` ever be empty?

File: README.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (liyanhui1228) [reply README.rst]: The names are slightly different. Java use the name `exporter` for reporters. And they call different  propagation like `BinaryFormat`. For samplers the names are same. And we use the same names as PHP did.

File: trace/opencensus/trace/request_tracer.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+
+class RequestTracer(object):
+    """The RequestTracer is for tracing a request for web applications.
+
+    :type span_context: :class:`~opencensus.trace.span_context.SpanContext`
+    :param span_context: SpanContext encapsulates the current context within
+                         the request's trace.
+
+    :type sampler: :class:`type`
```
[the remark this answers is not in the document]
Comment (liyanhui1228) [reply trace/opencensus/trace/request_tracer.py]: Fixed.

File: trace/opencensus/trace/request_tracer.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            propagator = google_cloud_format.GoogleCloudFormatPropagator()
+
+        self.span_context = span_context
+        self.sampler = sampler
+        self.reporter = reporter
+        self.propagator = propagator
+        self.tracer = self.get_tracer()
+
+    def get_enabled(self):
+        """Determine whether to sample this request or not.
```
[the remark this answers is not in the document]
Comment (liyanhui1228) [reply trace/opencensus/trace/request_tracer.py]: Changed to `should_sample`.

File: trace/opencensus/trace/request_tracer.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        If the context forces not tracing, just set enabled to False.
+        Else follow the sampler.
+
+        :rtype: bool
+        :returns: Whether to trace the request or not.
+        """
+        if self.span_context.enabled is False:
```
[the remark this answers is not in the document]
Comment (liyanhui1228) [reply trace/opencensus/trace/request_tracer.py]: Good call, done.

File: trace/opencensus/trace/request_tracer.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        :returns: The TraceSpan object.
+        """
+        return self.tracer.span(name)
+
+    def start_span(self, name=None):
+        return self.tracer.start_span(name)
+
+    def end_span(self):
+        """End a span. Remove the span from the span stack, and update the
+        span_id in TraceContext as the current span_id which is the peek
```
[the remark this answers is not in the document]
Comment (liyanhui1228) [reply trace/opencensus/trace/request_tracer.py]: Should be the top span in the span_stack, which will be the parent span of the spans created after it.

File: trace/opencensus/trace/tracer/base.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        :param label_value: Label value.
+        """
+        raise NotImplementedError
+
+    def get_trace_json(self):
+        """Get the JSON format trace."""
+        raise NotImplementedError
+
+    def list_collected_spans(self):
+        """List collected spans."""
```
[the remark this answers is not in the document]
Comment (liyanhui1228) [reply trace/opencensus/trace/tracer/base.py]: Added.

File: trace/opencensus/trace/tracer/context_tracer.py (hunk 2 above)
Comment (liyanhui1228) [reply trace/opencensus/trace/tracer/context_tracer.py]: Yes, I added the check for it.

File: README.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (chingor13) [reply README.rst]: PHP isn't committed to those names. It probably makes sense to use the same names across all the languages. We can sync up with the Java folks about this at the Tuesday meeting and get it written into a spec somewhere.

[the remark this answers is not in the document]
Comment (liyanhui1228) [reply README.rst]: Yes sounds good to me, we need to sync up the names, also default value for some parameters like sampling rate.

**Comment by liyanhui1228** [comment]
@duggelz Ready for another review, thanks!

**Pull request merged**